### ISSUE TITLE
FR: Use branch drawing symbols for `jj log` for supported terminals

### ISSUE THREAD

**Issue by mitchellh** [martinvonz/jj#4904]
**Is your feature request related to a problem? Please describe.**

Not really, just a nice to have 😄 

**Describe the solution you'd like**

Kitty recently added support for [branch drawing symbols](https://github.com/kovidgoyal/kitty/pull/7681).
I also added support for this in [my terminal](https://mitchellh.com/ghostty).
Issues are open in [other terminals](https://github.com/wez/wezterm/issues/6328).

I'd love to see either an option to use these or some terminal detection to use these characters. 

**Additional context**
Screenshot from the PR:

![image](https://github.com/user-attachments/assets/38dc969c-fa7e-4132-9777-8d5933b3f119)

**Comment by martinvonz** [comment]
We use the `sapling-renderdag` crate for drawing the graph. @quark-zju, what do you think? Should we file a corresponding bug in the Sapling repo for this? I imagine it would be implemented as another drawing style (in addition to the ones you/we [already support](https://martinvonz.github.io/jj/latest/config/#graph-style)).

**Comment by quark-zju** [comment]
Maybe. It sounds like it needs more information than the existing renders to disambiguate. We might have similar logic in Javascript used in ISL. So maybe we can backport some logic from ISL Javascript to Rust.

**Comment by mitchellh** [comment]
Here is a screenshot of what this looks like in practice with a real git log:

![image](https://github.com/user-attachments/assets/2d0ba12f-0794-43f8-8025-782c74102f2f)

**Comment by thoughtpolice** [comment]
Is autodetection necessary? I feel like that's an endless rat's nest of issues, and depending on the installed font, time of day, and shirt color you're wearing it can very easily end up looking ugly for the user in combination with the graph nodes. This became apparently when we added custom `log_node` symbols earlier this year; at the start we wanted to just choose better node symbols, after a few days of discussion and people trying things, there were lots of edge cases where things looked bad, and there seemed to be little hope in expecting it to "just work". We instead went the route of making `log_node` customizable via templates, which was a much bigger change, but now users can choose any node symbols they wanted.

Would it be acceptable to just customize the set of curved symbols using something similar to the way we allow people to customize `log_node`? Last I looked at `sapling-renderdag`, it seemed like adding a custom set of curved symbols would not be that hard, nor change the surface area of the API too much.

Note that colouring the graph lines themselves is something we have talked about, but is a bigger change (and specific to us and doesn't require renderdag cooperation.)

**Comment by mitchellh** [comment]
> Is autodetection necessary? I feel like that's an endless rat's nest of issues, and depending on the installed font, time of day, and shirt color you're wearing it can very easily end up looking ugly for the user in combination with the graph nodes.

No, but I'll note that these symbols are provided directly by the terminal emulators rasterized perfectly for the grid size. They aren't provided by fonts, which I agree is a mess. So my recommendation towards auto detection was to specifically detect terminals that do this. 

> Would it be acceptable to just customize the set of curved symbols using a method similar to customizing log_node? 

That's fine with me, I'd just love a way to use them. It'd be nice if they looked beautiful out of the box but if I can set some config to use them I'm happy to do that. 

> Note that colouring the graph lines themselves is something we have talked about, but is a bigger change and not specific to this (nor should it require sapling-renderdag help.)

Indeed, I just wanted to show how they behaved with regards to standards SGR foreground attributes.

**Comment by joshka** [comment]
> No, but I'll note that these symbols are provided directly by the terminal emulators rasterized perfectly for the grid size. They aren't provided by fonts, which I agree is a mess. So my recommendation towards auto detection was to specifically detect terminals that do this.

I wonder if adding these to Nerdfonts would be worth considering. Even though the font versions don't get rasterized perfectly, having them in the right place in the font will make support much more obvious to the other terminals that haven't done that step yet. Added an issue to track this in https://github.com/rbong/flog-symbols/issues/3

**Comment by eproxus** [comment]
Is there a way to use these with a custom template? I cannot find a way to customize the default `log_node` template to actually output these as unicode. Pasting the actual Unicode character just prints the literal code as an ASCII representation (e.g. `<U+F5FA>`), and there doesn't seem to be a way to insert unicode code points using e.g. `\uF5FA`. I'm using Ghostty and `printf "\uF5FA"` works so it may be a Jujutsu issue?

**Comment by josephlou5** [comment]
> Is there a way to use these with a custom template? I cannot find a way to customize the default `log_node` template to actually output these as unicode. Pasting the actual Unicode character just prints the literal code as an ASCII representation (e.g. `<U+F5FA>`), and there doesn't seem to be a way to insert unicode code points using e.g. `\uF5FA`. I'm using Ghostty and `printf "\uF5FA"` works so it may be a Jujutsu issue?

It's not a Jujutsu issue; it's a `less` issue. By default jj uses `less -FRX` as the [pager](https://docs.jj-vcs.dev/latest/config/#pager), and `-F`/`--quit-if-one-screen` causes the pager to quit if the output is less than one full screen. However, the output still goes through `less`, which, at least on my machine, does not play nicely with Unicode symbols. The workaround for this was `export LESSUTFCHARDEF="E000-F8FF:p,F0000-FFFFD:p,100000-10FFFD:p"` to support UTF-8 characters, and then my custom log nodes started working.
